docs/proposals/201909_thanos_functional_sharding.md

```diff
@@ -0,0 +1,153 @@
+---
+title: Thanos Sharding for Long Term Retention Storage
+type: proposal
+menu: proposals
+status: proposed
+owner: bwplotka
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/1034 (main)
+* https://github.com/thanos-io/thanos/pull/1245 (attempt on compactor)
+* https://github.com/thanos-io/thanos/pull/1059 (attempt on store)
+* https://github.com/thanos-io/thanos/issues/1318 (subpaths, multitenancy)
+* Store issues because of large bucket: 
+  * https://github.com/thanos-io/thanos/issues/814
+  * https://github.com/thanos-io/thanos/issues/1455
+
+## Summary
+
+This document describes the motivation and design of the sharding of Thanos components in terms of operations against object storage.
+Additionally we touch smarter pre-filtering of shards on Querier.
+
+## Motivation
+
+Currently all components that reads from object store assumes that all the operations and functionality should be done based 
+on **all** the available blocks available in the certain bucket's root directory. 
+
+This is in most cases is totally fine, however with time and allowance of storing blocks from multiple `Sources` into the same bucket, 
+the number of objects in bucket can grow drastically.
+
+This means that with time you might want to scale out certain components e.g:
+
+* Compactor: Larger number of objects does not matter much, however compactor has to scale (CPU, network) with number of Sources pushing blocks to the object storage. 
+If you have multiple sources handled by the same compactor, with slower network and CPU you might not compact/downsample quick enough to cope with incoming blocks.
+    * This happens a lot if compactor is not deployed for longer periods and has to quickly catch up with large number of blocks (e.g couple of months).  
```

Comment:
I would rearrange this sentence to something like:

This happens a lot no compactor is deployed for longer periods and thus has to quickly catch up with large number of blocks (e.g couple of months).